### PR TITLE
feat: add groth16 proving

### PR DESCRIPTION
**Pull request by ibmp33**
- [ ] TODO: There are some problems with the api interface. I will complete it tomorrow.

### CODE REVIEW

File: plonky/src/groth16.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::bellman_ce::bn256::Bn256;
+    use crate::{circom_circuit::CircomCircuit, plonk, reader};
+    const CIRCUIT_FILE: &'static str =
+        concat!(env!("CARGO_MANIFEST_DIR"), "/../test/mycircuit.r1cs");
```
Comment (eigmax) [inline plonky/src/groth16.rs:71]: We already have a test r1cs: test/multiplier.r1cs.

File: plonky/src/groth16.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
+use crate::bellman_ce::groth16::{
+    create_random_proof, generate_random_parameters, prepare_verifying_key, verify_proof,
+    Parameters, PreparedVerifyingKey, Proof, VerifyingKey,
+};
+use crate::bellman_ce::Circuit;
+use crate::errors::{EigenError, Result};
```
Comment (eigmax) [inline plonky/src/groth16.rs:6]: Unused import: EigenError

File: plonky/src/groth16.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            witness: None,
+            wire_mapping: None,
+            aux_offset: plonk::AUX_OFFSET,
+        };
+        let mut rng = rand::thread_rng();
+        let params = Groth16::circuit_specific_setup(circuit.clone(), &mut rng)?;
+
+        let proof = Groth16::prove(&params.0, circuit.clone(), &mut rng)?;
+
+        // let verified = Groth16::verify_with_processed_vk(&params.1, &inputs, &proof)?;
```
Comment (eigmax) [inline plonky/src/groth16.rs:85]: Can not be verified?

File: plonky/src/groth16.rs (hunk 2 above)
Comment (eigmax) [reply plonky/src/groth16.rs:6]: Make sure there is no warning at all.

File: plonky/src/groth16.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    use crate::{circom_circuit::CircomCircuit, plonk, reader};
+    const CIRCUIT_FILE: &'static str =
+        concat!(env!("CARGO_MANIFEST_DIR"), "/../test/mycircuit.r1cs");
+    #[test]
+    fn groth16_proof() -> Result<()> {
+        let circuit: CircomCircuit<Bn256> = CircomCircuit {
+            r1cs: reader::load_r1cs(CIRCUIT_FILE),
+            witness: None,
+            wire_mapping: None,
+            aux_offset: plonk::AUX_OFFSET,
```
Comment (eigmax) [inline plonky/src/groth16.rs:78]: There is no need to setup aux_offset.  Just assign 0 to it.

Comment (eigmax) [reply plonky/src/groth16.rs:78]: The problem probably is raised here, cause the CircomCuit we made is for Plonk, not for Groth.  Meanwhile, you have to fill in the witness here by `calculate_witness`.

File: plonky/src/api.rs [hunk 5]
```diff
@@ -226,3 +228,33 @@ pub fn generate_aggregation_verifier(
     verifier::recursive_plonk_verifier::create_verifier_contract_from_default_template(config, sol);
     Result::Ok(())
 }
+
+pub fn groth16_setup(circuit_file: &String, witness: &String, curve_type: &str) -> Result<()> {
```
Comment (ibmp33) [inline plonky/src/api.rs:232]: Does the setup function output json format? Should the proving key and verification key be separated or should they be put together?

Comment (eigmax) [reply plonky/src/api.rs:232]: Be separated. VK will be distributed to the verifier.

File: plonky/src/api.rs [hunk 6]
```diff
@@ -226,3 +228,33 @@ pub fn generate_aggregation_verifier(
...
+    let mut rng = rand::thread_rng();
+    match curve_type {
+        "bn256" => {
+            let circuit: CircomCircuit<Bn256> = CircomCircuit {
+                r1cs: reader::load_r1cs(circuit_file),
+                witness: Some(reader::load_witness_from_file::<Bn256>(witness)),
+                wire_mapping: None,
+                aux_offset: 0,
+            };
+            Groth16::circuit_specific_setup(circuit, &mut rng);
```
Comment (eigmax) [inline plonky/src/api.rs:242]: The result, which may be an error, is not captured.

File: plonky/src/groth16.rs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::bellman_ce::bn256::Bn256;
+    use crate::circom_circuit::CircomCircuit;
+    use crate::reader;
+    const CIRCUIT_FILE: &'static str =
```
Comment (eigmax) [inline plonky/src/groth16.rs:85]: Do more tests to cover complex circuits by zkit

File: plonky/src/api.rs (hunk 6 above)
Comment (eigmax) [reply plonky/src/api.rs:242]: I think we may need to support bls12381 in plonk too. 

So we can merge the setup of plonk and groth16.  This is not first priority, we can do it after we finish all functional tests.

File: plonky/src/groth16.rs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        circuit: Self::Circuit,
+        rng: &mut R,
+    ) -> Result<(Self::ProvingKey, Self::PreparedVerificationKey)> {
+        #[cfg(debug_assertions)]
+        let t = std::time::Instant::now();
+
+        let pk = generate_random_parameters::<E, Self::Circuit, R>(circuit, rng)?;
+        let pvk = prepare_verifying_key(&pk.vk);
+
+        #[cfg(debug_assertions)]
```
Comment (eigmax) [inline plonky/src/groth16.rs:31]: I think this timer print is not only for debugging, we also need to figure out the time cost for release.

Comment (eigmax) [reply plonky/src/groth16.rs:31]: So remove the debug cfg here

File: groth16/src/api.rs [hunk 9]
```diff
@@ -0,0 +1,55 @@
+use crate::bellman_ce::pairing::{bls12_381::Bls12, bn256::Bn256};
+use algebraic::errors::{EigenError, Result};
+use algebraic::witness::{load_input_for_witness, WitnessCalculator};
+use algebraic::circom_circuit::CircomCircuit;
+use algebraic::reader::load_r1cs;
+use crate::groth16::Groth16;
+use crate::snark::SNARK;
+
+// TODO plz move this function into zkit and remove this file
```
Comment (eigmax) [inline groth16/src/api.rs:9]: Plz take a look at this TODO. @ibmp33

Comment (eigmax) [reply groth16/src/api.rs:9]: expose groth16 interface to zkit

File: algebraic/src/lib.rs
Comment (ChengYueJia) [inline algebraic/src/lib.rs:1]: The circom and r1cs related files(mod) don't belong to algebraic. They'd better in dsl.

File: groth16/src/snark.rs [hunk 10]
```diff
@@ -0,0 +1,36 @@
+use algebraic::errors::Result;
+use franklin_crypto::bellman::pairing::ff::PrimeField;
```
Comment (ChengYueJia) [inline groth16/src/snark.rs:2]: We don't have to import franklin_crypto, as its an implement of `plonk`. 

For here, we  can use the pairing crate instead `franklin` one.

File: groth16/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,48 @@
+[package]
+name = "groth16"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+byteorder = "1"
+franklin-crypto = { git = "https://github.com/matter-labs/franklin-crypto", branch = "beta", features = [ "plonk" ], version = "0.0.5"}
```
Comment (ChengYueJia) [inline groth16/Cargo.toml:10]: We don't have to import franklin_crypto, as its an implement of `plonk`. 

For here, we  can use the `bellman_ce`, `pairing` crate instead `franklin`.

Comment (eigmax) [reply groth16/Cargo.toml:10]: Fixed by cargo udeps

Comment (eigmax) [reply groth16/Cargo.toml:10]: Another thought is we have a GPU frankin crypto. Can you plz create an issue to discuess this?

File: groth16/src/groth16.rs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+use franklin_crypto::bellman::pairing::Engine;
+use rand::Rng;
+use std::marker::PhantomData;
+
+pub struct Groth16<E: Engine, C: Circuit<E>> {
+    _engine: PhantomData<E>,
+    _circuit: PhantomData<C>,
+}
+
+impl<E: Engine, C: Circuit<E>> SNARK<E::Fr> for Groth16<E, C> {
```
Comment (ChengYueJia) [inline groth16/src/groth16.rs:13]: As the groth16 and plonky are seperated into two crates, seems this trait should be remove.

File: groth16/src/snark.rs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+use rand::Rng;
+/// The basic functionality for a SNARK.
+pub trait SNARK<F: PrimeField> {
```
Comment (ChengYueJia) [inline groth16/src/snark.rs:5]: As the groth16 and plonky are seperated into two crates, seems this trait should be remove.

File: groth16/Cargo.toml (hunk 11 above)
Comment (ChengYueJia) [reply groth16/Cargo.toml:10]: Sure. see #129

Comment (eigmax) [reply groth16/Cargo.toml:10]: it supports, but we have not done the integration test for a while.

File: algebraic/src/lib.rs
Comment (eigmax) [reply algebraic/src/lib.rs:1]: Will keep track of this in https://github.com/0xEigenLabs/eigen-zkvm/issues/130

File: groth16/src/snark.rs (hunk 13 above)
Comment (eigmax) [reply groth16/src/snark.rs:5]: Done

File: zkit/src/groth16_api.rs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+                    }
+                })
+                .collect::<Vec<_>>();
+            let circuit = create_circuit_from_file::<Bn256>(circuit_file, Some(w));
+            let inputs = circuit.get_public_inputs().unwrap();
+            let proof = Groth16::prove(&pk, circuit.clone(), &mut rng)?;
+            let writer = std::fs::File::create(proof_file)?;
+            proof.write(writer)?;
+            let mut writer1 = std::fs::File::create(public_input_file)?;
+            write_fr_vec(&inputs, &mut writer1)?;
```
Comment (eigmax) [inline zkit/src/groth16_api.rs:83]: Should write a json file. Same for proof.

Comment (eigmax) [reply zkit/src/groth16_api.rs:83]: Keep track in https://github.com/0xEigenLabs/eigen-zkvm/issues/131

**Pull request merged**